### ISSUE TITLE
Find a way to remove IndexSearcher#search(Query query, CollectorOwner collectorOwner) before 10.0

### ISSUE THREAD

**Issue by gsmiller** [apache/lucene#13671]
### Description

GH#13568 introduced a new public method to `IndexSearcher` along with a new concept of a `CollectorOwner`. This was done to allow drill-sideways to use the new sandbox faceting functionality, but as @javanna (and others) have mentioned, it doesn't feel like an ideal end-state. Let's either, 1) find a way to remove this new API, or 2) find a way to converge the concepts of `CollectorManager` and `CollectorOwner`. Let's find a solution before the 10.0 release.

In the worst-case, let's revert the `IndexSearcher` and drill-sideways changes when 10.0 releases then iteratively find a new path forward? The sandbox faceting work can remain, it just wouldn't (yet) be usable with drill-sideways.

**Comment by epotyom** [comment]
The main reason for adding `CollectorOwner` class was to allow different drill-sideways dimensions to use `CollectorManager`s with different Collector `C` and result `T` types. Except for rolling back IndexSearcher and DrillSideways changes, I see three paths forward for this issue:

**1st option**: Use the same `C` and `T` types for all drill-sideways dimensions. It's likely that most users don't need different types for dimensions. And the ones that do need it can use `MultiCollectorManager` and cast `Object` result to the right type. See also this comment https://github.com/apache/lucene/pull/13568#issuecomment-2239124290 and response https://github.com/apache/lucene/pull/13568#issuecomment-2239235073

Pros: we can rollback `IndexSearcher` changes that add new public `search` method.

**2nd option** is to change `CollectorManager` to keep the list of `Collector`s. This requires making converting it from interface to class, or maintain the list in every interface implementation. I believe both solutions are not ideal, they are not backwards-compatible and not convenient for users. E.g. if we convert it to class, one won't be able to create a CollectorManager that also extends some other class.

Pros: allow different `C` and `T` types per drill-sideways dimensions without casting and without separate CollectorOwner class.

**3rd option**: Make CollectorOwner less of a separate entity. E.g. we can make it nested static class inside `CollectorManager` interface, and add default method to the interface to wrap. Something like:

```
public interface CollectorManager<C extends Collector, T> {
  // existing methods
  C newCollector() throws IOException;  
  T reduce(Collection<C> collectors) throws IOException;

  // new method to wrap this CollectorManager 
  default Container<C, T> getContainer() {
    return new Container<>(this);
  }

  final static class Container<C, T> {
    // all CollectorOwner methods
  }
}
```

Pros: allow different `C` and `T` types per drill-sideways dimension without casting; both IndexSearcher and DrillSideways(Query) don't have to manage lists of `Collector`s themselves.

I personally prefer 3rd option, but I suppose 1st option is the least controversial?

Links to related conversations in the original PR:
- https://github.com/apache/lucene/pull/13568#discussion_r1675802255
- https://github.com/apache/lucene/pull/13568#discussion_r1681255620

**Comment by gsmiller** [comment]
Thanks @epotyom for the detailed thoughts! If I'm understanding this correctly, it sounds like "option 1" would let us remove the need for `CollectorOwner` along with the new API in `IndexSearcher` while still supporting drill-sideways use with the new sandbox faceting implementation, but the same aggregations would need to be done across all dimensions? One thing I like about that option is that it separates out the concerns of, 1) making sure drill-sideways can work with the new faceting implementation, and 2) extending the capabilities of drill-sideways further to allow for different aggregations for certain dimensions (which I agree sounds useful, but also pretty uncommon). I think options 2 and 3 are both things we should consider more as a follow-up, but I think it could take some time to get it right and build consensus, which I'm not sure we have ahead of a planned 10.0 release. What do you think?

**Comment by javanna** [comment]
I am also fine with option 1) .  I'd be happy to remove the additional methods recently added to IndexSearcher.

I have seen similar problems in the past around the collector manager design, and how decoupled the `newCollector` and `reduce` methods are, so I would also be open to changes to that design. I find that resorting to `MultiCollectorManager` is not great either, because of all the casting needed. We could also introduce a more generic CollectorManager wrapper (similar to CollectorOwner that though extends CollectorManager), which DrillSideways and potentially additional consumers can use in the future. Maybe overkill though, this is a change that we can always make in the future in that it is not breaking.

**Comment by gsmiller** [comment]
Cool, sounds like we've got a good plan. I'll just add that I've been noodling on the idea for a while of seeing if we could actually get rid of `DrillSideways` and fold what it does into `IndexSearcher` in a way that's not specific to faceting. One use-case I've been playing with is the idea of generically being able to produce multiple match sets during a single query evaluation where each match set is a "near miss" as drill-sideways defines it. This is exactly what `DrillSideways` does, but it's tightly coupled to faceting. If we're trying to make drill-sideways more generic to work with any `CollectorManager`, I'd like to see if we could bake this into `IndexSearcher`. Maybe we could explore this with option 2 or 3 as a follow-up? I'll create an issue as soon as I have time to describe the use-case in more detail with some examples, etc. to motivate it.

In the meantime though, as far as moving forward with Option 1 goes, I think it's a little more than just pulling out the `IndexSearcher` API right @epotyom? I think we need a new method in `DrillSideways` that lets users pass in `CollectorManager`s for the sideways dims (since right now it assumes you want facet collectors)? Maybe we add a method that takes in a `Supplier<CollectorManager>` or something for the sideways dims?

**Comment by epotyom** [comment]
@gsmiller , there is new method

```java
public void DrillSideways#search(
  DrillDownQuery query,
  CollectorOwner<?, ?> drillDownCollectorOwner,
  List<CollectorOwner<?, ?>> drillSidewaysCollectorOwners)
```

([link](https://github.com/apache/lucene/blob/0f5359a2ac6687d466c71e5bf5fa8138a9c07934/lucene/facet/src/java/org/apache/lucene/facet/DrillSideways.java#L434C1-L437C63)) that I believe does what you describe - it takes `CollectorOwner` for drill down and for each drill sideways dimension. It's up to user to decide what they want to collect for each dimension, e.g. top hits, doc ID sets for the main `facets` module (`FacetsCollectorManager`), or facets using the new sandbox facets module. Using wildcard types for `CollectorOwner`s allows doing something like collecting top hits for one drill-sideways dimension, but collect facets for other dimension. So yeah, we will kind of loose this flexibility with the Option 1; but there is `MultiCollectorManager` workaround.

I'm working on changing this method to something like:

```java
public <C extends Collector, T, K extends Collector, R> Result<T, R> DrillSideways#search(
  DrillDownQuery query,
  CollectorManager<C, T> drillDownCollectorManagers,
  List<CollectorManager<K, R>> drillSidewaysCollectorManagers) {...}

// We need to add another result class that doesn't depend on Facets
public record Result<T, R>  (T drillDownResult, List<R> drillSidewaysResults) {}
```

I believe changing it will be more or less straight forward.

**Comment by mikemccand** [comment]
I added `blocker` label to this and removed `type:enhancement` (it seemed super weird to have an enhancement blocker!).

**Comment by gsmiller** [comment]
@epotyom right makes sense. I should have been a little more specific. When I said "new method" I was thinking along the same lines of changing the recently-added method to not need `CollectorOwner`.

The response types for drill-sideways are a bit of a mess at the moment as well IMO. Maybe your new response type could be general enough that we could move to using it in more places as a follow-up?

**Comment by epotyom** [comment]
Draft PR for this issue: https://github.com/apache/lucene/pull/13702 , 

------

@gsmiller ,

>  I'd like to see if we could bake this into IndexSearcher.
>  ...
> The response types for drill-sideways are a bit of a mess at the moment as well IMO. Maybe your new response type could be general enough that we could move to using it in more places as a follow-up?

I see what you mean now. Hmm, so the only difference between `DrillSideways#search(DrillDownQuery, CollectorManager, List<CollectorManager>)` method we are adding/changing https://github.com/apache/lucene/pull/13702 and `IndexSearcher#search(Query, CollectorManager)` that the latter takes only one CollectorManager. Maybe the missing piece here is a `CollectorManager` for drill sideways? Something like:

```
public class DrillSidewaysCollectorManager<C extends Collector, T, K extends Collector, R>  
        implements CollectorManager<DrillSidewaysCollector<C, K>, DrillSidewaysCollectorManager.Result<T, R>> {  
  
    @Override  
    public DrillSidewaysCollector<C, K> newCollector() throws IOException {  
        return new DrillSidewaysCollector<>(...);  
    }  
  
    @Override  
    public Result<T, R> reduce(Collection<DrillSidewaysCollector> collectors) throws IOException {  
        // TK: reduce every dimension
    }  
  
    public record Result<T, R>(T drillDownResult, List<R> drillSidewaysResults) {}  
}
```

It will of cause require other changes for example:
1. Merge `DrillDownQuery` and `DrillSidewaysQuery` classes to have `Query` implementation that can be created by user directly and that creates Weight -> Score, etc 
	1. Alternatively, we can try returning `DrillSidewaysQuery` when `DrillDownQuery#rewrite` is called?
2. Cast leaf collector to the future `DrillSidewaysLeafCollector` in `DrillSidewaysScorer` to handle exact match/near misses.
	1. Ideally we should avoid casting, but I'm not sure if there is better way to do it? We could try moving `DrillSidewaysScorer` logic to `DrillSidewaysLeafCollector`, but it seems challenging/not possible to move all scorer optimizations to `collect` method

We would also loose parallel drill-sideways `searchConcurrently` , but I'm not sure if anyone uses it, I'd expect that inter-segment parallelism + `DrillSidewaysScorer` is better for both throughput and latency?

**Comment by gsmiller** [comment]
Resolving this out since we merged and back-ported #13702. Thanks @epotyom !
